macros/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+    let f: ItemFn = syn::parse(input).expect("`#[pre_init]` must be applied to a function");
+
+    // check the function signature
+    assert!(
+        f.constness.is_none()
+            && f.vis == Visibility::Inherited
+            && f.unsafety.is_some()
+            && f.abi.is_none()
+            && f.decl.inputs.is_empty()
+            && f.decl.generics.params.is_empty()
+            && f.decl.generics.where_clause.is_none()
+            && f.decl.variadic.is_none()
+            && match f.decl.output {
+                ReturnType::Default => true,
+                ReturnType::Type(_, ref ty) => match **ty {
+                    Type::Tuple(ref tuple) => tuple.elems.is_empty(),
+                    _ => false,
+                },
+            },
+        "`#[pre_init]` function must have signature `unsafe fn()`"
+    );
+
+    assert_eq!(
+        args.to_string(),
+        "",
+        "`pre_init` attribute must have no arguments"
+    );
+
+    // XXX should we blacklist other attributes?
+    let attrs = f.attrs;
+    let ident = f.ident;
+    let block = f.block;
+
+    quote!(
+        #[export_name = "__pre_init"]
+        #(#attrs)*
+        pub unsafe fn #ident() #block
+    ).into()
+}
+
+/// Attribute to place a function in RAM
+///
+/// This attribute can *not* be used on generic functions.
+///
+/// # Examples
+///
+/// You can put plain functions in RAM
+///
+/// ```
+/// # use cortex_m_rt_macros::ramfunc;
+/// #[ramfunc]
+/// fn computation_heavy_function(data: &[u8]) -> u32 {
+///     // do something here
+/// #   0
+/// }
+/// # fn main() {}
+/// ```
+///
+/// And you can also put exception handlers in RAM.
+///
+/// ```
+/// # use cortex_m_rt_macros::{exception, ramfunc};
+/// #[exception(SysTick)]
+/// #[ramfunc]
+/// fn handler() {
+///     // do something here
+/// }
+/// # fn main() {}
+/// ```
+#[proc_macro_attribute]
+pub fn ramfunc(args: TokenStream, input: TokenStream) -> TokenStream {
+    assert_eq!(
+        args.to_string(),
+        "",
+        "`ramfunc` attribute must have no arguments"
+    );
+
+    let f: ItemFn = syn::parse(input).expect("`#[ramfunc]` must be applied to a function");
+
+    assert!(
+        f.decl.generics.params.is_empty(),
+        "`#[ramfunc]` can't be applied to generic functions"
+    );
+    // because then all instantiations of the function would end in the same linker section and the
+    // linker won't be able to GC the unused ones (read comment below for more details)
+
+    // NOTE we want to put *each* function into a *different* section so the linker is able to GC
+    // unused functions. Unlike the compiler we don't know the crate name or which module this
+    // function is in so we can't use a mangled version of the path to the function -- we don't know
+    // its full path! Instead we'll use a random string for the section name of each function.
+    let mut rng = rand::thread_rng();
+    let hash = (0..16)
```

Comment:
Using a RNG in a procedural macro doesn't fell quite right to me but I don't see any other option to generate a unique section name from only the AST of function. Also, I have no idea if the RNG negatively impacts incremental compilation. I don't think it affects the reproducibility of the builds because the random section names won't make it to the final binary -- the linker merge all of them and puts them under `.ramfunc`.